sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -161,33 +161,38 @@ case class Not(child: Expression)
        true
   """)
 // scalastyle:on line.size.limit
-case class In(value: Expression, list: Seq[Expression]) extends Predicate {
+case class In(values: Seq[Expression], list: Seq[Expression]) extends Predicate {
```

Comment:
can we have an analyzer rule to deal with `In(CreateStruct(...), ListQuery(...))`, to unpack the `CreateStruct`, or pack the `ListQuery`? Then we don't need to change `In`.

Reply:
I don't think so, as the `value` can be replaced later by other rules. So we do need to have a `Seq[Expression]` here, instead of a single expression. Another possible option which I haven't checked, but I think it may be feasible is to create a new kind of `Expression` (eg. `InValues`) we can use only for this specific case. What do you think?

Reply:
+1 on `InValues`. Maybe call it `InSubquery`

Reply:
it is not a subquery, this is the "left part" of IN, so I don't really agree on `InSubquery`, but if you have another suggestion I am happy to follow it. Thanks.

Reply:
I mean `case class InSubquery(values: Seq[Expression], subquery: ListSubquery)`, it's not just the left part.

Reply:
Well, that would really work. We have the same problem also in the case we don't have a subquery. Eg. from Postgres:
```
mgaido=# select 1 from (select (1, 'a') as col1) tab1 where col1 in ((1, 'a'));
ERROR:  could not identify an equality operator for type unknown
mgaido=# select 1 from (select (1, 'a') as col1) tab1 where (1, 'a') in ((1, 'a'));
 ?column? 
----------
        1
(1 row)
```

Reply:
ah it's getting interesting. I tried `=`
```
cloud0fan=# select 1 from (select (1, 'a') as col1) tab1 where col1 = (1, 'a');
 ERROR:  could not identify an equality operator for type unknown
cloud0fan=# select 1 from (select 1 as col1, 'a' as col2) tab1 where (col1, col2) = (1, 'a');
 ?column? 
----------
        1
(1 row)
```
I'm wondering if there is any other special rules for `(...)`

Reply:
I am not sure which Postgres version you're using but that seems a bug. I am using version 10 and it works as expected, ie.:

```
mgaido=# select 1 from (select (1, 'a') as col1) tab1 where col1 = (1, 'a');
ERROR:  could not identify an equality operator for type unknown
mgaido=# select 1 from (select 1 as col1, 'a' as col2) tab1 where (col1, col2) = (1, 'a');
 ?column? 
----------
        1
(1 row)
```

Anyway, in Presto, both work:

```
presto> select 1 from (select (1, 'a') as col1) tab1 where col1 in ((1, 'a'));
 _col0 
-------
     1 
(1 row)

Query 20180628_163550_00000_gkbmf, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
0:01 [0 rows, 0B] [0 rows/s, 0B/s]

presto> select 1 from (select (1, 'a') as col1) tab1 where (1, 'a') in ((1, 'a'));
 _col0 
-------
     1 
(1 row)
Query 20180628_163600_00001_gkbmf, FINISHED, 1 node
Splits: 17 total, 17 done (100.00%)
0:00 [0 rows, 0B] [0 rows/s, 0B/s]

presto> select 1 from (select (1, 'a') as col1) tab1 where col1 in (select (1, 'a'));
 _col0 
-------
     1 
(1 row)

Query 20180628_163607_00002_gkbmf, FINISHED, 1 node
Splits: 83 total, 83 done (100.00%)
0:01 [0 rows, 0B] [0 rows/s, 0B/s]

presto> select 1 from (select (1, 'a') as col1) tab1 where (1, 'a') in (select (1, 'a'));
 _col0 
-------
     1 
(1 row)
```

Reply:
Hive behaves as Postgres
```
0: jdbc:hive2://aaa> select 1 from (select struct(1, 'a') as col1) tab1 where col1 in ((1, 'a'));
Error: Error while compiling statement: FAILED: ParseException line 1:74 cannot recognize input near ')' '<EOF>' '<EOF>' in expression specification (state=42000,code=40000)
0: jdbc:hive2://aaa> select 1 from (select struct(1, 'a') as col1) tab1 where (1, 'a') in ((1, 'a'));
+------+--+
| _c0  |
+------+--+
| 1    |
+------+--+
1 row selected (0.074 seconds)
```

So I do believe this is the right direction.

Reply:
The postgres result you posted for `=` is same as mine, isn't it?

Reply:
oh, sorry, you're right. I got confused with Presto's results. Sorry.

Reply:
Anyway, I think right behavior is the one which both Postgres and Hive have (and it is also the same of Oracle/MySQL, in which we don't have structs). What do you think?

Reply:
I agree we should treat `(...)` specially if it's in front of `In`, but I'm wondering if we need to do the same thing for `=`.

Reply:
I am not sure. The behavior when comparing structs in not uniform among different DBs. Hive doesn't allow `=` on structs. Postgres and Presto does, but their behavior with nulls is not consistent and it is different from ours. In particular, comparing a struct containing a `null` returns `null` on Postgres and causes an exception in Presto (we return `false` instead). This is causing another problem which has been reported in another JIRA for which we can return results different from Postgres and Oracle (SPARK-24395).

Reply:
For this specific case, instead, I'll update this PR creating the new ad-hoc expression for the values in front of IN if you agree, as we have to deal not only with the subquery case. What do you think?

Reply:
SGTM